alchemiscale_fah/compute/client.py

```diff
@@ -0,0 +1,608 @@
+        )
+
+    @classmethod
+    def write_key(cls, key, key_file):
+        pem = key.private_bytes(
+            encoding=serialization.Encoding.PEM,
+            format=serialization.PrivateFormat.TraditionalOpenSSL,
+            encryption_algorithm=serialization.NoEncryption(),
+        )
+
+        with open(key_file, "wb") as f:
+            f.write(pem)
+
+    @classmethod
+    def generate_csr(cls, key, csr_file, email):
+        """Generate certificate signing request (CSR) using private key.
+
+        It is necessary to create a CSR and present this to an AS in order to
+        receive a valid certificate. The CSR will be written in PEM format.
+
+        Parameters
+        ----------
+        key
+            RSA private key, as returned by `create_key` or `read_key`.
+        csr_file
+            Path in which to place CSR generated by this method.
+        email
+            Email address used to authenticate with the AS.
+
+        """
+        cn = x509.NameAttribute(NameOID.COMMON_NAME, email)
+        csr = x509.CertificateSigningRequestBuilder()
+        csr = csr.subject_name(x509.Name([cn]))
+        csr = csr.sign(key, hashes.SHA256())
+
+        with open(csr_file, "wb") as f:
+            f.write(csr.public_bytes(serialization.Encoding.PEM))
+
+    @staticmethod
+    def read_csr_pem(csr_file):
+        with open(csr_file, "r") as f:
+            pem = f.read()
+
+        return pem
+
+    def _check_status(self, r):
+        if r.status_code != 200:
+            raise Exception("Request failed with %d: %s" % (r.status_code, r.text))
+
+    def _get(self, api_url, endpoint, **params):
+        url = urljoin(api_url, endpoint)
+        r = requests.get(url, cert=self.cert, params=params, verify=self.verify)
+        self._check_status(r)
+        return r.json()
+
+    def _put(self, api_url, endpoint, data=None):
+        url = urljoin(api_url, endpoint)
+        r = requests.put(url, json=data, cert=self.cert, verify=self.verify)
+        self._check_status(r)
+
+    def _post(self, api_url, endpoint, data=None):
+        url = urljoin(api_url, endpoint)
+        r = requests.post(url, json=data, cert=self.cert, verify=self.verify)
+        self._check_status(r)
+        return r.content
+
+    def _delete(self, api_url, endpoint):
+        url = urljoin(api_url, endpoint)
+        r = requests.delete(url, cert=self.cert, verify=self.verify)
+        self._check_status(r)
+
+    def _upload(self, api_url, endpoint, filename):
+        url = urljoin(api_url, endpoint)
+        with open(filename, "rb") as f:
+            r = requests.put(url, data=f, cert=self.cert, verify=self.verify)
+            self._check_status(r)
+
+    def _upload_bytes(self, api_url, endpoint, bytedata):
+        url = urljoin(api_url, endpoint)
+        r = requests.put(url, data=bytedata, cert=self.cert, verify=self.verify)
+        self._check_status(r)
+
+    def _download(self, api_url, endpoint, filename):
```

Comment:
NIT: Two semi-different meanings of `download` here, one to file, while `_download_bytes` returns in-memory

Reply:
Correct, this is the distinction here. I think I'll stick with this as the established pattern since we use it so pervasively here.